daemon/oci_windows.go

```diff
@@ -249,7 +250,41 @@ func (daemon *Daemon) createSpecWindowsFields(c *container.Container, s *specs.S
 			return errors.New("createSpecWindowsFields: BaseFS of container " + c.ID + " is unexpectedly empty")
 		}
 
-		s.Root.Path = c.BaseFS // This is not set for Hyper-V containers
+		if daemon.UsesSnapshotter() {
+			// HACK: daemon.Mount() for the snapshotter actually mounts the filesystem to the host
+			// using containerd/mount.All and BaseFS is the directory where this is mounted.
+			// This is consistent with Linux.
+			// For the windowsfilter graphdriver, the underlying Get() call does not actually mount
+			// the filesystem to a path, and BaseFS is the Volume GUID of the prepared/activated
+			// filesystem.
+			// hcsshim is only prepared to handle the latter format, of course, and having the mount
+			// active on the host might cause other problems...
+			// ALTERNATIVES: Either shift containerd image-store mode to use WithSnapshot instead of
+			// locally mounting the container's root volume; or make daemon.Mount() consistent with the
+			// graphdriver on Windows.
+			// The latter is tempting, but because Unmount relies on getting the mount-point back,
+			// it's actually conceptually a hassle.
+
+			mappings, err := bindfilter.GetBindMappings(c.BaseFS)
+			if err != nil {
+				return errors.Wrapf(err, "createSpecWindowsFields: Failed to get bind-mappings of BaseFS of container %s", c.ID)
+			}
+			log.G(context.TODO()).Errorf("GetBindMappings: %v", mappings)
+			if len(mappings) == 0 {
+				return fmt.Errorf("GetBindMappings for container %s has unexpectedly %d mappings", c.ID, len(mappings))
+			}
+			// HACK/TODO: It seems we get the same mapping multiple times. Not sure if bug or expected.
+			if len(mappings) > 1 {
+				log.G(context.TODO()).Errorf("GetBindMappings for container %s has unexpectedly %d mappings", c.ID, len(mappings))
+			}
+			mapping := mappings[0]
+			if len(mapping.Targets) != 1 {
+				return fmt.Errorf("GetBindMappings for container %s mount point %s has unexpected %d mappings", c.ID, mapping.MountPoint, len(mapping.Targets))
+			}
+			s.Root.Path = mapping.Targets[0]
```

Comment:
I was thinking about this, and I can't imagine any feasible API change in containerd to unbreak the encapsulation here, short of just lifting this block of code into `containerd/mount/mount_windows.go`. The main issue is that the volume name we need is created as part of the mounting process, but is not an end-point, and the mount API is _very_ simple, there no place to return anything except the target mount-point (or a replacement of it if necessary).

The graphdriver, AFAICT, avoids this because when it needs to operate on the volume directly, i.e. copy, it directly uses the volume path as the root path, it never mounts to the local filesystem first, so the volume path _is_ the end-point.

Perhaps containerd should have worked that way too, but its API would have made that difficult, as `Mount` takes a target path we'd have to ignore, and `Unmount` takes _only_ the target path but the underlying Windows code does not provide a way to find the necessary data for unmounting given only a volume path. (We avoid this when mounting to the filesystem by writing the necessary data to an ADS for the target path)

Hmm. I just realised that an alternative implementation would be to read the layer data out of the ADS file and use `hcs.GetLayerMountPath` to get the same data. That's still poking at internals, but is probably _more accurate_ and also has a better chance of being uplifted into containerd.

This all goes away if we use WithSnapshot, of course.

Reply:
An advantage of the `hcsshim.GetLayerMountPath` approach is that it won't pick up random bindfilter mounts if you somehow give it the wrong directory.

On the other hand, if we do implement real bind-mounts on Windows for some reason, then `hcsshim.GetLayerMountPath` would be wrong. But if that happens, it could be dealt with at the time.

So yeah, I think my preferred solution to this is a new function that reads the ADS file and uses `hcsshim.GetLayerMountPath`, and then propose upstreaming that into containerd.

Reply:
I rewrote this commit to break the logic out into a function, and then added https://github.com/moby/moby/pull/46539/commits/51f1243b57ffab849f15c5b15e546076653804f0 which replaces the logic with `hcsshim.GetLayerMountPath`. I think the latter is a vast improvement (no weird behaviours or likely edge cases), but welcome opinions before I squash it and unflag it as a hack.